enhancements/baremetal/baremetal-config-raid-and-bios.md

```diff
@@ -0,0 +1,133 @@
+---
+
+# Config RAID and BIOS for Baremetal IPI deployments
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes to add the support of RAID and BIOS configuration for baremetal IPI deployments.
+
+## Motivation
+
+With the new feature of [RAID configuration](https://github.com/metal3-io/baremetal-operator/pull/292) and the upcoming feature of [BIOS configuration](https://github.com/metal3-io/baremetal-operator/pull/302), openshift already has or will have the ability to configure RAID and BIOS for baremetal.
+
+However, these features only work with a running cluster, users cannot configure desired RAID and BIOS during IPI deployments.
+
+### Goals
+
+- Allow users to configure RAID and BIOS in ***install-config.yaml***.
+- IPI deployments can handle the configuration of RAID and BIOS.
+
+### Non-Goals
+
+- Add new or vendor specific options regarding RAID and BIOS.
+
+## Proposal
+
+The configuration of RAID and BIOS is finally handed over to Ironic during the IPI deployments.
+1. Add two new fields *raid* and *firmware* to *platform.baremetal.hosts* in the [install-config.yaml](https://github.com/openshift/installer/blob/master/data/data/install.openshift.io_installconfigs.yaml) file.
+2. Openshift Installer process the field contents into manifests.
+3. [Terraform-provider-ironic](https://github.com/openshift-metal3/terraform-provider-ironic) gets the
+configuration of RAID and BIOS for master nodes and then passes it to Ironic([Baremetal Operator(BMO)](https://github.com/metal3-io/baremetal-operator)
+has supported the process of RAID and BIOS for worker nodes).
+
+### User Stories
+
+With the addition of this feature, the users can configure RAID and BIOS in IPI deployments.
+
+### Implementation Details/Notes/Constraints
+
+#### Add two fields
+
+New *platform.baremetal.hosts* fields called *raid* and *firmware*.
+
+```yaml
+platform:
+  baremetal:
+    ...
+    hosts:
+      - name: master-0
+        role: master
+        bmc:
+          address: IPAddress
+          username: UserName
+          password: PassWord
+        bootMACAddress: MACAddress
+        raid: RAIDConfig
+        firmware: BIOSConfig
+```
+
+RAID feature has been implemented in **BMO**, so the *raid* field here is the same as the [BMH](https://github.com/metal3-io/baremetal-operator/blob/399f5ef7ee3831014c1425250bc4fa49641a8709/config/crd/bases/metal3.io_baremetalhosts.yaml).
+The *firmware* field is the same as the *spec.firmware* field in **BMH** which is been advancing by [#302](https://github.com/metal3-io/baremetal-operator/pull/302).
```

Comment:
We typically try to keep the API of install-config small, and limited to things that almost every deployment needs to manage. I don't think that applies to RAID or BIOS settings, even for bare metal deployments.

The installer runs in multiple phases. Typically a user will run `create cluster` and get all of the phases, but they can also run `create manifests` and `create cluster` separately. Between those steps, they can add and modify the manifests used to create the cluster.

Is it possible to pass the RAID and BIOS settings this way, instead of through install-config? The user would run `create manifests` and then add or replace the `BareMetalHost` manifests created by the installer. The `create manifests` step would read the new files and the interface to terraform in the installer code would need to understand the BIOS and RAID fields.

Reply:
> Is it possible to pass the RAID and BIOS settings this way, instead of through install-config?

Yes, it is possible. We used this way as an [alternative](https://github.com/openshift/enhancements/pull/791/files#diff-b06bf476cab4c460016a94b59469061638772af0097367a394ad9a23fee21c83R123-R125) before. 

> The `create manifests` step would read the new files and the interface to terraform in the installer code would need to understand the BIOS and RAID fields.

I think you may mean the `create cluster` step here. We would modify the installer code to make the interface to terraform understand the BIOS and RAID fields.

Reply:
Modifying the manifests seems like a reasonable compromise, provided we can consume that data via terraform in a reasonably clean way in the installer code.

One thing to mention is that the BMH resources will still get persisted in the cluster, so the modified master BMH resources will have the BIOS/RAID config, but also be `externallyProvisioned: true`

Provided the BMO can handle that, it's probably OK, but I wanted to note it's a corner case since BMH resources probably won't normally be defined that way.

Reply:
@hardys Thanks for your review.

> the modified master BMH resources will have the BIOS/RAID config, but also be externallyProvisioned: true

This is indeed what you said. We plan to add this note to the proposal.

Reply:
+1 to using BMH API for BIOS, especially since we're going to have 2 slightly different approaches to BIOS settings.